File: hotel-mern/services/building.service.ts
import { Building, BuildingPopulated } from '../shared/models'
import { BuildingRepository } from '../shared/repositories'

export default class BuildingService extends BuildingRepository {
  async get(): Promise<Building[]> {
    return this.model.find({}).lean()
  }

  async getOne(_id: string): Promise<BuildingPopulated> {
    return this.model.findById(_id).populate('_rooms').lean()
  }

  async create(address: string): Promise<Building> {
    const building = await new this.model({ address })
    await building.save()
    return building
  }

  async change(_id: string, address: string): Promise<Building> {
    const building: Building = await this.model
      .findByIdAndUpdate(_id, {
        $set: { address },
      })
      .lean()
    building.address = address
    return building
  }

  async addRoom(buildingId: string, roomId: string) {
    await this.model.updateOne(
      { _id: buildingId },
      { $push: { _rooms: roomId } }
    )
  }

  async removeRoom(roomId: string) {
    await this.model.updateMany({}, { $pull: { _rooms: roomId } })
  }

  async delete(_id: string): Promise<string> {
    await this.model.findByIdAndRemove(_id)
    return _id
  }
}

File: hotel/backend/src/main/kotlin/loshica/api/hotel/services/BuildingService.kt
package loshica.api.hotel.services

import loshica.api.hotel.core.BaseService
import loshica.api.hotel.interfaces.IBuildingService
import loshica.api.hotel.models.Building
import loshica.api.hotel.models.Room
import loshica.api.hotel.repositories.BuildingRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class BuildingService(
    @Autowired override val repository: BuildingRepository
) : BaseService<Building, BuildingRepository>(repository), IBuildingService {

    override fun create(address: String): Building {
        val building = Building(address = address)
        repository.save(building)
        return building
    }

    override fun change(id: Int, address: String): Building {
        val building: Building = getOne(id)
        building.address = address
        repository.save(building)
        return building
    }

    override fun addRoom(room: Room) {
        room.building.let {
            it.rooms.add(room)
            repository.save(it)
        }
    }

    override fun removeRoom(room: Room) {
        room.building.let {
            it.rooms.remove(room)
            repository.save(it)
        }
    }
}
